torch/csrc/autograd/engine.cpp

```diff
@@ -82,106 +288,75 @@ auto Engine::backward(const variable_list& variables,
 
-    int size = grad_inputs.size();
-    for (int i = 0; i < size; ++i) {
-      auto& grad_input = grad_inputs[i];
-      auto& prev_fn = fn->previous_functions[i].first;
-      int output_nr = fn->previous_functions[i].second;
+  // Search the graph and find all stochastic functions. Append them to the queue.
+  find_stochastic_functions(creators, backward_task);
 
-      // null inputs have no previous_function and we skip them here
-      if (!prev_fn) {
-        continue;
-      }
+  if (!backward_task.node_requires_grad) {
+    throw std::runtime_error(
+      "there are no graph nodes that require computing gradients");
+  }
 
-      if (auto var = dynamic_cast<Variable*>(prev_fn.get())) {
-        if (var->requires_grad) {
-          var->backward(grad_input);
-        }
-        continue;
-      }
+  // Now compute the dependencies for each function which requires grad
+  compute_dependencies(std::move(creators), backward_task);
 
-      // Stochastic functions are placed in the ready queue by
-      // compute_dependencies, so we can skip them here.
-      if (prev_fn->is_stochastic || !prev_fn->requires_grad) {
-        continue;
-      }
+  // wait for all tasks to complete
+  backward_task.not_done.wait(lock, [&backward_task]{
+    return backward_task.outstanding_tasks.load() == 0;
+  });
 
-      // Check if the function is ready for backward
-      bool is_ready = false;
-      auto it = dependencies.find(prev_fn.get());
-      if (it == dependencies.end()) {
-        throw std::runtime_error("dependency not found");
-      } else if (--it->second == 0) {
-        dependencies.erase(it);
-        is_ready = true;
-      }
-
-      auto not_ready_it = not_ready.find(prev_fn.get());
-      if (is_ready) {
-        if (not_ready_it == not_ready.end()) {
-          // The function is ready and no buffers have been allocated for it
-          GradBuffer prev_buffer(prev_fn->num_outputs);
-          prev_buffer.addGrad(output_nr, std::move(grad_input));
-          ready.emplace_front(prev_fn, std::move(prev_buffer));
-        } else {
-          // The function is ready and it already has a buffer allocated.
-          auto prev_buffer = std::move(not_ready_it->second);
-          not_ready.erase(not_ready_it);
-          prev_buffer.addGrad(output_nr, std::move(grad_input));
-          ready.emplace_front(prev_fn, std::move(prev_buffer));
-        }
-      } else {
-        // Allocate a buffer if necessary and accumulate gradient
-        if (not_ready_it == not_ready.end()) {
-          GradBuffer prev_buffer(prev_fn->num_outputs);
-          prev_buffer.addGrad(output_nr, std::move(grad_input));
-          not_ready.emplace(prev_fn.get(), std::move(prev_buffer));
-        } else {
-          auto &prev_buffer = not_ready_it->second;
-          prev_buffer.addGrad(output_nr, std::move(grad_input));
-        }
-      }
-    }
+  // check for an exception while running backwards
+  if (backward_task.has_error.load()) {
```

Comment:
nit: `std::atomic` always puts a memory fece around reads, so there's no need to call `load()` explicitly. But we can leave that